modules/SofaMiscMapping/src/SofaMiscMapping/SubsetMultiMapping.inl

```diff
@@ -25,6 +25,7 @@
 
 #include <sofa/linearalgebra/EigenSparseMatrix.h>
 #include <sofa/core/MappingHelper.h>
+#include <SofaBaseMechanics/MechanicalObject.h>
```

Comment:
This is quite a big dependency. 
I never worked with only a Point Topology so I never thought about this scenario. If I got it, if you only have mechanicalObjects in your scene the mapping works fine because it resize the number of dof but when you have TopologyContainer, this one is not updated?
Does the resize come from a mechanical operation or is it due to a topological change?
All the current design has been done to update the topology first which warn all the components to update themself (using the callbacks).
The link/interaction between mechanicalObject and PointSetTopologyContainer is not clear neither for me. Just to understand what is a subsetMultiMapping compare to a SubsetMapping?

Reply:
Let's say you have 2 mechanical objects `mstateA` and `mstateB`.
You create a third object `mstateC`, and it "contains" the points 12 and 65 of `mstateA`, and the points 8 and 32 of `mstateB`.
`mstateC` is a "fusion" of subsets of `mstateA` and `mstateB`. This is done with a SubsetMultiMapping.

Let's say you have a topological change in `mstateB`. The SubsetMultiMapping must have a callback registered to the topology of `mstateB`, right? In that case, I assumed the mapping must propagate this change to the context of `mstateC`, because `mstateC` depends on `mstateB`.

You should have a look on the scene examples/Components/topology/TopologicalModifiers/RemovingPointSubsetMultiMapping.scn I introduced in https://github.com/sofa-framework/sofa/pull/2729

Reply:
To reduce the dependency, I could add a virtual method to retrieve the topology in BaseState. The dependency would be on BaseState.

Reply:
The change looks a bit weird to me because it seems to esthablish an  implicit connexion between the subset multimapping and a topology going thru mechanical object.

Reply:
You asked about "how to apply the mapping to both the geometry and the topology? 
And pointed that  "usualy it is the topology which apply a change to the mechanical state and not the other way around". A possible direction to look at, is to keep the "usual way". If a mapping needs to change the topology then it has a link to the topology and probably in-heriting to TopologicalMapping (or engine). Doing so is would be more explicit clearly showing that the mapping is changing the topology. 

More generally this rise a question me about the link between MechanicalObject and topologies.  
Fundamentally I was thinking about MechanicalObject as a "topology free object". Just a container of numerical values and derivatives without any topological or geometric interpretation. It was the role of forcefield to build upon a topological and geometrical description some way to fill's the dofs' content or a visual model to interpret values as 3d position to render. With this view in mind I was more expecting that it was the FF role to react to topology change and having mechanical object out of the picture.

Reply:
That makes sense. But I don't know well `TopologicalMapping`. @epernod do you think it makes sense to have `SubsetMultiMapping` deriving from `TopologicalMapping`?
Just a warning here: `SubsetMultiMapping` is probably not the only mechanical mapping to have such an issue of propagation of topological changes. It means such mappings will be both mechanical AND topological mappings at once.
Another issue to look at is that `TopologicalMapping` takes a single input and a single output, while `SubsetMultiMapping` can takes multiple input.